### DISCUSSION TITLE
error: no matching function for call to 'FEProblemBase::assembly(THREAD_ID&, int)

### DISCUSSION THREAD

**Discussion by batodon** [idaholab/moose#22996] [Q&A General] [answered]
Hello,
I’m assigning a THREAD_ID typedef to some variables (_curr_node and _curr_element), but I keep getting the following no matching function call when I try to compile. Could you please help?

```
/home/batodon/MOOSE/projects/ergs/src/materials/PFEMBase.C:94:38: error: no matching function for call to 'FEProblemBase::assembly(THREAD_ID&, int)'
   94 |     _curr_node(problem.assembly(tid,0).node()),
      |                                      ^
In file included from /home/batodon/MOOSE/projects/moose/framework/build/header_symlinks/UserObjectInterface.h:14,
                 from /home/batodon/MOOSE/projects/moose/framework/build/header_symlinks/MaterialBase.h:21,
                 from /home/batodon/MOOSE/projects/moose/framework/build/header_symlinks/Material.h:13,
                 from /home/batodon/MOOSE/projects/moose/modules/porous_flow/build/header_symlinks/PorousFlowMaterial.h:12,
                 from /home/batodon/MOOSE/projects/moose/modules/porous_flow/build/header_symlinks/PorousFlowMaterialVectorBase.h:12,
                 from /home/batodon/MOOSE/projects/moose/modules/porous_flow/build/header_symlinks/PorousFlowPermeabilityBase.h:12,
                 from /home/batodon/MOOSE/projects/ergs/build/header_symlinks/PFEMBase.h:3,
                 from /home/batodon/MOOSE/projects/ergs/src/materials/PFEMBase.C:1,
                 from /home/batodon/MOOSE/projects/ergs/build/unity_src/materials_Unity.C:2:
/home/batodon/MOOSE/projects/moose/framework/build/header_symlinks/FEProblemBase.h:330:14: note: candidate: 'virtual Assembly& FEProblemBase::assembly(THREAD_ID)'
  330 |   Assembly & assembly(THREAD_ID tid) override
      |              ^~~~~~~~
/home/batodon/MOOSE/projects/moose/framework/build/header_symlinks/FEProblemBase.h:330:14: note:   candidate expects 1 argument, 2 provided
/home/batodon/MOOSE/projects/moose/framework/build/header_symlinks/FEProblemBase.h:335:20: note: candidate: 'virtual const Assembly& FEProblemBase::assembly(THREAD_ID) const'
  335 |   const Assembly & assembly(THREAD_ID tid) const override
      |                    ^~~~~~~~
/home/batodon/MOOSE/projects/moose/framework/build/header_symlinks/FEProblemBase.h:335:20: note:   candidate expects 1 argument, 2 provided
In file included from /home/batodon/MOOSE/projects/ergs/build/unity_src/materials_Unity.C:2:
/home/batodon/MOOSE/projects/ergs/src/materials/PFEMBase.C:95:41: error: no matching function for call to 'FEProblemBase::assembly(THREAD_ID&, int)'
   95 |     _curr_element(problem.assembly(tid,0).elem()),
```

Here is my declaration:
```
	  FEProblemBase & problem;
	  THREAD_ID tid;
```
And part of my destructor:

```
    problem(*getCheckedPointerParam<FEProblemBase *>("_fe_problem_base")),
    tid(getParam<THREAD_ID>("_tid")),
    _curr_node(problem.assembly(tid,0).node()),
    _curr_element(problem.assembly(tid,0).elem()),
```
The problem seems to come from the arguments to _curr_node and _curr_element. Thanks.

**Comment by GiudGiud** [comment, marked as answer]
I think you just need problem.assembly(tid), no 0
What does that return ?

**Reply by batodon** [reply]
Thanks, @GiudGiud. I already tried it. It compiled but I had a segmentation fault when I run my input file.

**Reply by GiudGiud** [reply]
it might not come from that though

can you run this in a debugger
https://mooseframework.inl.gov/application_development/debugging.html

**Reply by batodon** [reply]
Hi @GiudGiud. Any help? I hit a 'virtual memory exhausted' when building in a` dbg` mode:
```
{standard input}: virtual memory exhausted: Cannot allocate memory
virtual memory exhausted: Cannot allocate memory
virtual memory exhausted: Cannot allocate memory
Assembler messages:
{standard input}:709029: Fatal error: bfd_make_empty_symbol: memory exhausted
{standard input}:709029: Fatal error: /home/batodon/MOOSE/projects/moose/modules/tensor_mechanics/build/unity_src/.libs/userobjects_Unity.x86_64-conda-linux-gnu.dbg.o: memory exhausted
virtual memory exhausted: Cannot allocate memory
virtual memory exhausted: Cannot allocate memory
make: *** [/home/batodon/MOOSE/projects/moose/framework/build.mk:144: /home/batodon/MOOSE/projects/moose/modules/tensor_mechanics/build/unity_src/materials_lagrangian_Unity.x86_64-conda-linux-gnu.dbg.lo] Error 1
make: *** Waiting for unfinished jobs....
make: *** [/home/batodon/MOOSE/projects/moose/framework/build.mk:144: /home/batodon/MOOSE/projects/moose/modules/tensor_mechanics/build/unity_src/userobjects_Unity.x86_64-conda-linux-gnu.dbg.lo] Error 1
make: *** [/home/batodon/MOOSE/projects/moose/framework/build.mk:144: /home/batodon/MOOSE/projects/moose/modules/tensor_mechanics/build/unity_src/materials_Unity.x86_64-conda-linux-gnu.dbg.lo] Error 1
make: *** [/home/batodon/MOOSE/projects/moose/framework/build.mk:144: /home/batodon/MOOSE/projects/moose/modules/tensor_mechanics/build/unity_src/materials_cohesive_zone_model_Unity.x86_64-conda-linux-gnu.dbg.lo] Error 1
make: *** [/home/batodon/MOOSE/projects/moose/framework/build.mk:144: /home/batodon/MOOSE/projects/moose/modules/tensor_mechanics/build/unity_src/materials_crystal_plasticity_Unity.x86_64-conda-linux-gnu.dbg.lo] Error 1
make: *** [/home/batodon/MOOSE/projects/moose/framework/build.mk:144: /home/batodon/MOOSE/projects/moose/modules/tensor_mechanics/build/unity_src/bcs_Unity.x86_64-conda-linux-gnu.dbg.lo] Error 1
 ```

**Reply by GiudGiud** [reply]
can you build with a single process?
METHOD=dbg make -j 1

**Reply by batodon** [reply]
Not quite sure if I'm getting the correct backtrace, but here is the error from debugging. I can't find the location (to either the header/src file) for my problem.

```
*** ERROR ***

Stack frames: 7
0: libMesh::print_trace(std::ostream&)
1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
2: void mooseError<char const*>(char const*&&)
3: MooseApp::run()
4: main
5: __libc_start_main
6: /home/batodon/MOOSE/projects/ergs/ergs-dbg(+0x3079) [0x555555557079]

Thread 1 "ergs-dbg" hit Breakpoint 1, 0x00007fffea061cc0 in PMPI_Abort () from /home/batodon/mambaforge3/envs/moose/lib/libmpi.so.12
(gdb) bt
#0  0x00007fffea061cc0 in PMPI_Abort () from /home/batodon/mambaforge3/envs/moose/lib/libmpi.so.12
#1  0x00007ffff4a75bab in moose::internal::mooseErrorRaw (msg=..., prefix=...)
    at /home/batodon/MOOSE/projects/moose/framework/src/base/MooseError.C:83
#2  0x00007ffff5f21ff5 in mooseError<char const*> () at /home/batodon/MOOSE/projects/moose/framework/build/header_symlinks/MooseError.h:258
#3  0x00007ffff4a165a9 in MooseApp::run (this=0x5555557474a0) at /home/batodon/MOOSE/projects/moose/framework/src/base/MooseApp.C:1406
#4  0x0000555555557739 in main (argc=3, argv=0x7fffffffbed8) at /home/batodon/MOOSE/projects/ergs/src/main.C:33
```

**Reply by GiudGiud** [reply]
This error with no message is often a libmesh error.

The backtrace shows line 1406 in the mooseApp file
What is it for your current moose version?

For me it s the errorCheck, but for you it might be something else

**Reply by batodon** [reply]
Thanks @GiudGiud. This is what I have on line 1406 of mooseApp.C file:

  `mooseError(err.what());`

**Reply by GiudGiud** [reply]
ok so an initialSetup() is failing with a libmesh_error and it s being caught there/

You ll need to set a breakpoint on libmesh_error, then generate a backtrace there

**Reply by batodon** [reply]
Sorry, I got  `No stack. ` as  backtrace after setting a breakpoint on libmesh_error.

**Reply by GiudGiud** [reply]
you ll need to catch the breakpoint to have a stack.
no stock means it ran to completion

**Reply by batodon** [reply]
Sorry, I'm probably not getting something right. This is the command line I'm using to set the breakpoint on libmesh_error `b libmesh_error`.  if it ran to completion, does it mean no error detected?

**Reply by GiudGiud** [reply]
Did you add a new object?
This is pointing you to the intialSetup routine


I m thinking we re not going to be able to catch a breakpoint on a libmesh error. This is because you are using the conda installed libmesh and not a debugging version.

you could try to install libmesh from source and re run the debugger 

or you could paste your object here see if someone jumps to mind

**Reply by batodon** [reply]
Thanks, @GiudGiud. Yup, I added some new objects to my class. Here they are:

```
PFEMBase::~PFEMBase() {}

void
PFEMBase::setRandomResetFrequency(ExecFlagType exec_flag)
{
  _reset_on = exec_flag;
  _ri_problem.registerRandomInterface(*this, _ri_name);
}

void
PFEMBase::setRandomDataPointer(RandomData * random_data)
{
  _random_data = random_data;
  _generator = &_random_data->getGenerator();
}

unsigned int
PFEMBase::getSeed(std::size_t id)
{
  mooseAssert(_random_data, "RandomData object is NULL!");

  return _random_data->getSeed(id);
}

Real
PFEMBase::getRandmField() const
{
  mooseAssert(_generator, "Random Generator is NULL, did you call setRandomResetFrequency()?");

  dof_id_type id;
  if (_is_nodal)
    id = _curr_node->id();
  else
    id = _curr_element->id();

  return _generator->rand(id);
}
```

Their corresponding declaration, constructors and destructors are as follows:  
```
protected:
	~PFEMBase();

	void setRandomResetFrequency(ExecFlagType exec_flag);
	Real getRandmField() const;
	unsigned int getSeed(std::size_t id);

	void setRandomDataPointer(RandomData* random_data);

	unsigned int getMasterSeed() const { return _master_seed; }
	bool isNodal() const { return _is_nodal; }
	ExecFlagType getResetOnTime() const { return _reset_on; }

	  FEProblemBase & problem;
	  THREAD_ID tid;
	  bool is_nodal;
	  
private:
	  RandomData * _random_data;
	  mutable MooseRandom * _generator;
	
	  FEProblemBase & _ri_problem;
	  const std::string _ri_name;
	
	  unsigned int _master_seed;
	  bool _is_nodal;
	  ExecFlagType _reset_on;
	
	  const Node* const& _curr_node;
	  const Elem* const& _curr_element;
```

```
    params.addParam<unsigned int>("seed", 0, "The seed for the master random number generator");
    params.addParamNamesToGroup("seed", "Advanced");
```

```
  _random_data(nullptr),
  _generator(nullptr),
  _ri_problem(problem),
  _ri_name(parameters.get<std::string>("_object_name")),
  _master_seed(parameters.get<unsigned int>("seed")),
  _is_nodal(is_nodal),
  _reset_on(EXEC_LINEAR),
  problem(*getCheckedPointerParam<FEProblemBase *>("_fe_problem_base")),
  tid(getParam<THREAD_ID>("_tid")),
  _curr_node(problem.assembly(tid).node()),
  _curr_element(problem.assembly(tid).elem())
```

**Reply by GiudGiud** [reply]
Tid is not a parameter. 

_tid should be an existing class attribute which you can use for the construction. 

If you are not using threads you can just use 0 for the thread id

**Reply by GiudGiud** [reply]
oh my bad it seems using a tid parameter is quite common actually. It must be something else then

What is PFEMBase supposed to be ? An AuxKernel?

**Reply by batodon** [reply]
Thanks, @GiudGiud. I tried different scenarios, including the one you recommended (i.e., using 0 instead for the thread id). I still get a ‘segmentation fault.’ PFEMBase is a material class, not an auxKernel. The added objects are supposed to be functions which can be called within the PFEMBase to compute a property. Yup, tid is also used here: https://mooseframework.inl.gov/docs/doxygen/moose/RandomInterface_8C_source.html

**Reply by batodon** [reply]
Any help on this,pls? I hit a snag!

**Reply by GiudGiud** [reply]
We ll need a backtrace
Around line 1406 you have a try / catch statement in MooseApp.C

Remove it and recompile
This will let you error at the right location, and get a proper backtrace

**Reply by batodon** [reply]
Thanks, @GiudGiud. I got the following bt after hitting a breakpoint on `MPI_Abort`. Same as before, there is no exact error pointing to a location in my src/header files.

```
(gdb) bt
#0  0x00007fffea048cc0 in PMPI_Abort () from /home/batodon/mambaforge3/envs/moose/lib/libmpi.so.12
#1  0x00007ffff4a5c6b7 in moose::internal::mooseErrorRaw (msg=..., prefix=...) at /home/batodon/MOOSE/projects/moose/framework/src/base/MooseError.C:83
#2  0x00007ffff6da2ca2 in mooseError<char const (&) [54]> () at /home/batodon/MOOSE/projects/moose/framework/build/header_symlinks/MooseError.h:258
#3  0x00007ffff49fa227 in MooseApp::executeExecutioner (this=0x55555574f520) at /home/batodon/MOOSE/projects/moose/framework/src/base/MooseApp.C:1101
#4  0x00007ffff49fce84 in MooseApp::run (this=0x55555574f520) at /home/batodon/MOOSE/projects/moose/framework/src/base/MooseApp.C:1403
#5  0x0000555555557739 in main (argc=3, argv=0x7fffffffbed8) at /home/batodon/MOOSE/projects/ergs/src/main.C:33
```

**Reply by GiudGiud** [reply]
which line is that?
#3  0x00007ffff49fa227 in MooseApp::executeExecutioner (this=0x55555574f520) at /home/batodon/MOOSE/projects/moose/framework/src/base/MooseApp.C:1101

**Reply by batodon** [reply]
Sorry, I have this  on line 1101 instead. 

```
  mooseError("No executioner was specified (go fix your input file)");
```

**Reply by batodon** [reply]
But, I have an executioner in my input file:

```
[Executioner]
  type = Steady
  solve_type = Newton
  [TimeStepper]
    type = IterationAdaptiveDT
    dt = 1E4  # 1E3 1E4
    growth_factor = 1.2
    optimal_iterations = 10
  []
  nl_abs_tol = 1E-12
[]
```

**Reply by GiudGiud** [reply]
Can you turn on show_actions = true in the [Debug] block? See if there is something odd with when this object is added
I dont know why you are hitting this error at this point. I think you might want to upload your code to a github repo and we can see if we can take a look

**Reply by batodon** [reply]
Hello @GiudGiud, Thanks. Nothing interesting really happened when `show_actions = true`. The code is available at this repo https://github.com/iyevugah/ERGS/tree/main/src/materials.  Please, note that this problem stems from ticket #22971. A previous implementation (PFEMBase1) was successful, but after further querying the results, it became obvious that the field only had one random value. A second attempt to obtain the random fields (i.e., PFEMBase2) is the culprit here (i.e., yields segmentation fault). The input file is in the problem2 directory. Thanks a lot!

**Reply by batodon** [reply]
@GiudGiud, as an alternative, I’ve tried rotating the element id instead. For an annular model, I ended up with the left image below, which gave me a corresponding pore pressure field on the right.

![perm_pp](https://user-images.githubusercontent.com/113298585/209190582-8892f393-8592-47e1-ab8f-cc11fe485ceb.PNG)


From the benchmark study I’m working on, I should get a random field like the left image (shown below), with its corresponding pore pressure field on the right.  Please, note that while the pressure fields look similar, the permeability field was a bit off. Thanks.

![perm_pp_benchmark](https://user-images.githubusercontent.com/113298585/209190312-c6419035-498c-4362-ab9b-4206b118bb0f.PNG)

**Reply by GiudGiud** [reply]
Can you please give me the specs of what you need again?
3 fields, that are random spatially, and change randomly every time step?

**Reply by batodon** [reply]
Hi @GiudGiud, I’m looking to rotate each element separately based on their corresponding random values at the end of each simulation. Therefore, my rotation matrix could act on any random and spatially varied 'scalar' field to achieve this result. No, I only need 1 scalar field. Do you think 3 fields are needed? (I may be missing something). And this field shouldn’t change randomly every timestep; that will defeat the purpose of having a fixed random field. Please, let me know if you need any extra info that might help me. Thanks.

**Reply by batodon** [reply]
You can look at these random field values as random rotation angles passed as arguments to a rotation matrix. I hope this helps.

**Reply by GiudGiud** [reply]
I said 3 because usually you need 3 angles to fully define a rotation matrix (well for a general rotation. Around a fixed axis, 1 angle is fine)

If you only need one field, and constant in time, what was the issue with the field generated by the RandomIC (stored in an auxvariable) ?
Sorry if you already explained this

**Reply by batodon** [reply]
You’re right, @GiudGiud. I’m rotating 2 orthotropic axes. I have 2 rotation matrices, and I’m supposed to have 2 rotation angles (i.e., 2 random fields). But, since the rotation angles are similar, one rotation angle suffices. In the implementation, I assign this (one) rotation angle to two different variables so that each rotation matrix has an argument (this is inefficient coding on my part). 

The RandomIC gave me an initial condition: the same random value across the board. What I need is spatially different random values for all elements. The RandomInterface class (you rightly recommended) could do this. I followed the definition of some of its member functions to build my material class, but I’m hitting a segmentation fault. I don’t know if obtaining an initial condition through RandomIC from the input file into my material class (as the rotation angle) is possible (and could do the job). Is this something you’ll recommend? Sorry for making this a lengthy one. Thanks again.

**Reply by GiudGiud** [reply]
The randomIc should not be the same value across the board

Look at the input in test/tests/ics/random_ic_test
It generates this:
![image](https://user-images.githubusercontent.com/25037151/209246631-293362ae-3f88-4e3f-803b-329e336a3a5e.png)


You should be able to generate the two angles with two randomICs in two auxvariables

**Reply by batodon** [reply]
This is beautiful, @GiudGiud!!! What will I do without you? (lol). Now I’m trying to pass the two aux variables (i.e., the rotation field) as an argument to my rotation matrix in the src file. How can I do this? Is there an example moose file for that? You’re awesome!!! The best!

**Reply by GiudGiud** [reply]
you can look at the coupledForce kernel to see how to do that in a kernel / auxkernel

It s through the coupleable interface

**Reply by batodon** [reply]
Hello @GiudGiud, Here are my results, and they are not quite what I expected. The aux variable came out right (left image). When I passed it (as an argument) to the rotation matrix, I had the pressure field (right image ), which doesn't really look like what I expected. The pressure field looks more like it wasn’t rotated. I’m still working on it. But for now, it seems  my aux variable is not influencing the rotation. 

![rotxy_rotyz](https://user-images.githubusercontent.com/113298585/209263804-f8846d8a-0442-4c76-bf44-f15e7bc35e90.PNG)

**Reply by GiudGiud** [reply]
how do you retrieve and use the random field in the item? could you please link to that code

**Reply by GiudGiud** [reply]
my guess is that it wasnt stored as a reference

**Reply by batodon** [reply]
Hi @GiudGiud, I made it a const reference, as shown below for the declaration and the constructor: 
```
	const Real& _rotXY;
	const Real&  _rotYZ;
```

```
      params.addRequiredCoupledVar("rotation_angleXY", "The rotation field.");
      params.addRequiredCoupledVar("rotation_angleYZ", "The rotation field.");
```

```
               _rotXY(coupled("rotation_angleXY")),
                _rotYZ(coupled("rotation_angleYZ"))
```

The input file for the annular model is uploaded in the repo. For the annular input file, use PFEMBase and PFOrthoEM, instead. I've tried using `coupledGradient`, but that didn't work either.

**Reply by GiudGiud** [reply]
coupled is an integer, the index of the variable

you need to use coupledValue

**Reply by batodon** [reply]
I've got a compile-error after changing it to `coupledValue` and declaring it as `int` (even if I declare it as `real`): 

```
/home/batodon/MOOSE/projects/ergs/src/materials/PFEMBase.C:86:25: error: invalid initialization of reference of type 'const int&' from expression of type 'const VariableValue' {aka 'const MooseArray<double>'}
   86 |      _rotXY(coupledValue("rotation_angleXY")),
      |             ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
/home/batodon/MOOSE/projects/ergs/src/materials/PFEMBase.C:87:25: error: invalid initialization of reference of type 'const int&' from expression of type 'const VariableValue' {aka 'const MooseArray<double>'}
   87 |      _rotYZ(coupledValue("rotation_angleYZ"))
```

**Reply by GiudGiud** [reply]
the variable types are wrong.
this is no longer a real (in the header)
it should be a const VariableValue& _rotXY

**Reply by batodon** [reply]
Yeah, true. I declared `_rotXY` as a VariableValue, and I got a new compile-error:

```
/home/batodon/MOOSE/projects/ergs/src/materials/PFEMBase.C:145:33: error: cannot convert 'const VariableValue' {aka 'const MooseArray<double>'} to 'MooseADWrapper<double, false>' {aka 'double'} in assignment
  145 |            _randm_rad_xy[_qp] = _rotXY; // _current_elem->id(); // getRandomReal() * (_max - _min) + _min;
      |                                 ^~~~~~
      |                                 |
      |                                 const VariableValue {aka const MooseArray<double>}
/home/batodon/MOOSE/projects/ergs/src/materials/PFEMBase.C:146:33: error: cannot convert 'const VariableValue' {aka 'const MooseArray<double>'} to 'MooseADWrapper<double, false>' {aka 'double'} in assignment
  146 |            _randm_rad_yz[_qp] = _rotYZ; //_current_elem->id(); // getRandomReal() * (_max - _min) + _min;
      |                                 ^~~~~~
      |                                 |
      |                                 const VariableValue {aka const MooseArray<double>}
```
And when I declare `_rad_xy` as a VariableValue, I got a new error asking for `_rad_xy` to be declared in the rotation matrix `rotMat_xy`.

**Reply by GiudGiud** [reply]
yeah you need to take a good look at the kernel I mentioned. You can use the doxygen for help
https://mooseframework.inl.gov/docs/doxygen/moose/classCoupleable.html

**Reply by batodon** [reply]
Sure, I'll take a look at it and let you know shortly. Thanks again, @GiudGiud

**Reply by batodon** [reply]
Hello @GiudGiud. There is “`no matching function for the call to cos(VariableValue&).`” I’m guessing this error means that MOOSE can not compute the cosine of VariableValue&. If this is true, it defeats the purpose of using the aux variable as a rotation angle. Is there a way around this error?

**Reply by GiudGiud** [reply]
if _u is the variable value, then you need to index into the variableValue with the quadrature point

so cos(_u[_qp])

**Reply by batodon** [reply]
Eureka! I got it!, @GiudGiud. It ran, beautifully! 

![eureka](https://user-images.githubusercontent.com/113298585/209405644-a0cba08d-a14d-45d4-b945-55c4ce758e14.PNG)

**Reply by batodon** [reply]
Now, I can say my permeability material is fully working. I want to use this opportunity to thank you, @GiudGiud, for patiently guiding me exceptionally through this implementation. I also want to thank @cpgr for his guidance along the way. And to the Moose Team and everyone who contributed! Happy holidays!

**Reply by GiudGiud** [reply]
This is great news! Happy holidays to you too!
